flocker/node/agents/ebs.py

```diff
@@ -683,6 +701,87 @@ def _is_cluster_volume(cluster_id, ebs_volume):
     return False
 
 
+def _attach_volume_and_wait_for_device(
+    volume, attach_to,
+    attach_volume, detach_volume,
+    device, attach_attempt,
+    blockdevices,
+):
+    """
+    Attempt to attach an EBS volume to an EC2 instance and wait for the
+    corresponding OS device to become available.
+
+    :param BlockDeviceVolume volume: The Flocker representation of the volume
+        to attach.
+    :param unicode attach_to: The EC2 instance id to which to attach it.
+    :param attach_volume: A function like ``EC2Connection.attach_volume``.
+    :param attach_volume: A function like ``EC2Connection.detach_volume``.
+    :param unicode device: The OS device path to which to attach the device.
+    :param int attach_attempt: The number of this attempt to attach the volume
+        (counting from 1).  This is used to decide whether it's worth
+        continuing to retry or not.
+    :param list blockdevices: The OS device paths which are already present on
+        the system before this operation is attempted (primarily useful to make
+        testing easier).
+
+    :raise: Anything ``attach_volume`` can raise.  Or
+        ``AttachedUnexpectedDevice`` if the volume appears to become attached
+        to the wrong OS device file.
+
+    :return: ``True`` if the volume is attached and accessible via the expected
+        OS device file.  ``False`` if the attempt times out without succeeding.
```

Comment:
Hrm... seems more like, "``False`` if attach_volume fails with an InvalidParameterValue and attach_attempt was less than MAX_ATTACH_RETRIES."

This seems like a confusing interface to me. I think I would rather have the:
```
+        if e.code == u'InvalidParameterValue':
+            if attach_attempt == MAX_ATTACH_RETRIES:
+                raise
+            return False
+        else:
+            raise
```
block 1 layer up in the stack, or have attach_attempt be a boolean instead of a count, and be named something like return_false_instead_of_raise_invalid_parameter_value - or something more manageable.  To me those semantics indicate "this logic belongs a layer up".